File: actuar/src/loglogistic.c
</*  ===== actuar: an R package for Actuarial Science =====
 *
 *  Fonctions to compute density, cumulative distribution and quantile
 *  fonctions of the Pareto distribution, and to simulate random
 *  variates. See ../R/pareto.R for details.
 *
 *  AUTHORS: Mathieu Pigeon and Vincent Goulet <vincent.goulet@act.ulaval.ca>
 */

#include <R.h>
#include <Rmath.h>
#include "locale.h"

double dloglogistic(double x, double gamma, double scale, int give_log)
{
    if (!R_FINITE(gamma) ||
	!R_FINITE(scale) ||
	gamma <= 0.0 || 
	scale <= 0.0 || 
	x < 0.0) 
	error(_("invalid arguments"));
    
    return  give_log ?
	log(gamma) + gamma * (log(x) - log(scale)) - log(x) - 2.0 * log(1 + R_pow(x / scale, gamma)) :
	gamma * R_pow(x / scale, gamma)/(x * R_pow(1.0 + R_pow(x / scale, gamma), 2.0));
}

double rloglogistic(double gamma, double scale)
{	
    double a;
	
    if (!R_FINITE(gamma) ||
	!R_FINITE(scale) ||
	gamma <= 0.0 ||
	scale <= 0.0)
	error(_("invalid arguments"));

    a = unif_rand();

    return scale * R_pow(a / (1.0 - a), 1.0 / gamma);
}

File: actuar/R/loglogistic.R
### ===== actuar: an R package for Actuarial Science =====
###
### Definition of the {d,p,q,r}loglogistic functions. The loglogistic
### distribution used in these functions has cumulative distribution
### function
###
###   Pr[X <= x] = ((x/scale)^gamma/(1 + (x/scale)^gamma)), x > 0.
###
### See Appendix A of Klugman, Panjer & Willmot, Loss Models, Second
### Edition, Wiley, 2004.
###
### AUTHORS:  Mathieu Pigeon, Vincent Goulet <vincent.goulet@act.ulaval.ca>

dloglogistic <- function (x, gamma, scale, log)
  .External("do_dpq", "dloglogistic", x, gamma, scale, log)

## ploglogistic <- function(x, gamma, scale)
##     

## qloglogistic <- function(q, gamma, scale)
##    

rloglogistic <- function(n, gamma, scale)
    .External("do_random", "rloglogistic", n, gamma, scale)
